File: cantrip_app/ios/App/App/CantripInstallationStoragePlugin.swift
import Capacitor
import Foundation

@objc(CantripInstallationStoragePlugin)
public final class CantripInstallationStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CantripInstallationStoragePlugin"
    public let jsName = "CantripInstallationStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "status", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readCatalog", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "applyCatalogTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "replaceMissingKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inspectKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unwrapAccountMasterKey", returnType: CAPPluginReturnPromise)
    ]

    private let queue = DispatchQueue(label: "art.cantrip.installation-storage", qos: .userInitiated)
    private var storage: CantripInstallationStorage?

    @objc override public func load() {
        do {
            storage = try CantripInstallationStorage()
        } catch {
            storage = nil
        }
    }

    @objc func status(_ call: CAPPluginCall) {
        perform(call) { storage in storage.status() }
    }

    @objc func readCatalog(_ call: CAPPluginCall) {
        perform(call) { storage in try storage.readCatalog() }
    }

    @objc func applyCatalogTransaction(_ call: CAPPluginCall) {
        perform(call) { storage in
            guard let request = call.getObject("request") else {
                throw CantripNativeStorageError("installation-catalog-corrupt")
            }
            return try storage.applyCatalogTransaction(request)
        }
    }

    @objc func createKey(_ call: CAPPluginCall) {
        perform(call) { storage in
            guard let request = call.getObject("input") else {
                throw CantripNativeStorageError("native-device-key-invalid")
            }
            return try storage.createKey(request)
        }
    }

    @objc func replaceMissingKey(_ call: CAPPluginCall) {
        perform(call) { storage in
            guard let request = call.getObject("input") else {
                throw CantripNativeStorageError("native-device-key-invalid")
            }
            return try storage.replaceMissingKey(request)
        }
    }

    @objc func inspectKey(_ call: CAPPluginCall) {
        performOptional(call) { storage in
            guard let keyAlias = call.getString("keyAlias") else {
                throw CantripNativeStorageError("native-device-key-invalid")
            }
            return try storage.inspectKey(keyAlias)
        }
    }

    @objc func unwrapAccountMasterKey(_ call: CAPPluginCall) {
        queue.async { [weak self] in
            do {
                guard let storage = self?.storage else {
                    throw CantripNativeStorageError("native-key-store-unavailable")
                }
                guard let request = call.getObject("input") else {
                    throw CantripNativeStorageError("client-master-key-wrapper-invalid")
                }
                var plaintext = try storage.unwrapAccountMasterKey(request)
                defer { plaintext.resetBytes(in: 0..<plaintext.count) }
                call.resolve(["bytes": plaintext.map(Int.init)])
            } catch {
                self?.reject(call, error)
            }
        }
    }

    private func perform(_ call: CAPPluginCall, operation: @escaping (CantripInstallationStorage) throws -> [String: Any]) {
        queue.async { [weak self] in
            do {
                guard let storage = self?.storage else {
                    throw CantripNativeStorageError("native-key-store-unavailable")
                }
                call.resolve(try operation(storage))
            } catch {
                self?.reject(call, error)
            }
        }
    }

    private func performOptional(
        _ call: CAPPluginCall,
        operation: @escaping (CantripInstallationStorage) throws -> [String: Any]?
    ) {
        queue.async { [weak self] in
            do {
                guard let storage = self?.storage else {
                    throw CantripNativeStorageError("native-key-store-unavailable")
                }
                if let result = try operation(storage) {
                    call.resolve(result)
                } else {
                    call.resolve()
                }
            } catch {
                self?.reject(call, error)
            }
        }
    }

    private func reject(_ call: CAPPluginCall, _ error: Error) {
        let native = error as? CantripNativeStorageError
        let code = native?.code ?? "native-storage-task-failed"
        call.reject(code, code, error)
    }
}

File: cantrip_app/android/app/src/main/java/art/cantrip/CantripInstallationStoragePlugin.java
package art.cantrip;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CantripInstallationStorage")
public final class CantripInstallationStoragePlugin extends Plugin {

    private CantripInstallationStorage storage;

    @Override
    public void load() {
        storage = new CantripInstallationStorage(getContext().getApplicationContext());
    }

    @PluginMethod
    public void status(PluginCall call) {
        call.resolve(storage.status());
    }

    @PluginMethod
    public void readCatalog(PluginCall call) {
        resolve(call, () -> storage.readCatalog());
    }

    @PluginMethod
    public void applyCatalogTransaction(PluginCall call) {
        resolve(call, () -> {
            JSObject request = call.getObject("request");
            if (request == null) throw new CantripInstallationStorage.StorageException("installation-catalog-corrupt");
            return storage.applyCatalogTransaction(request);
        });
    }

    @PluginMethod
    public void createKey(PluginCall call) {
        resolve(call, () -> {
            JSObject input = call.getObject("input");
            if (input == null) throw new CantripInstallationStorage.StorageException("native-device-key-invalid");
            return storage.createKey(input);
        });
    }

    @PluginMethod
    public void replaceMissingKey(PluginCall call) {
        resolve(call, () -> {
            JSObject input = call.getObject("input");
            if (input == null) throw new CantripInstallationStorage.StorageException("native-device-key-invalid");
            return storage.replaceMissingKey(input);
        });
    }

    @PluginMethod
    public void inspectKey(PluginCall call) {
        try {
            String keyAlias = call.getString("keyAlias");
            if (keyAlias == null) throw new CantripInstallationStorage.StorageException("native-device-key-invalid");
            JSObject result = storage.inspectKey(keyAlias);
            if (result == null) call.resolve(); else call.resolve(result);
        } catch (CantripInstallationStorage.StorageException error) {
            call.reject(error.code, error.code, error);
        }
    }

    @PluginMethod
    public void unwrapAccountMasterKey(PluginCall call) {
        resolve(call, () -> {
            JSObject input = call.getObject("input");
            if (input == null) throw new CantripInstallationStorage.StorageException("client-master-key-wrapper-invalid");
            byte[] plaintext = storage.unwrapAccountMasterKey(input);
            try {
                JSArray bytes = new JSArray();
                for (byte value : plaintext) bytes.put(value & 0xff);
                return new JSObject().put("bytes", bytes);
            } finally {
                CantripHpke.clear(plaintext);
            }
        });
    }

    private interface Operation {
        JSObject run() throws CantripInstallationStorage.StorageException;
    }

    private void resolve(PluginCall call, Operation operation) {
        try {
            call.resolve(operation.run());
        } catch (CantripInstallationStorage.StorageException error) {
            call.reject(error.code, error.code, error);
        } catch (Exception error) {
            call.reject("native-storage-task-failed", "native-storage-task-failed", error);
        }
    }
}
